temporal/server.go

```diff
@@ -411,6 +416,75 @@ func verifyPersistenceCompatibleVersion(config config.Persistence, persistenceSe
 	return nil
 }
 
+// TODO: remove this func after 1.10 release
+func copyCustomSearchAttributesFromDynamicConfigToClusterMetadata(cfg *config.Config, persistenceServiceResolver resolver.ServiceResolver, logger log.Logger, dc *dynamicconfig.Collection) {
+	if !cfg.Persistence.IsAdvancedVisibilityConfigExist() {
+		logger.Debug("Advanced visibility is not configured. Search attributes migration is cancelled.")
+		return
+	}
+
+	advancedVisibilityDataStore, ok := cfg.Persistence.DataStores[cfg.Persistence.AdvancedVisibilityStore]
+	if !ok {
+		logger.Debug("Advanced visibility data store is not configured. Search attributes migration is cancelled.")
+		return
+	}
+
+	visibilityIndex := advancedVisibilityDataStore.ElasticSearch.GetVisibilityIndex()
+	if visibilityIndex == "" {
+		logger.Debug("Advanced visibility Elasticsearch index is not configured. Search attributes migration is cancelled.")
+		return
+	}
+
+	dcSearchAttributes, err := searchattribute.BuildTypeMap(dc.GetMapProperty(dynamicconfig.ValidSearchAttributes, map[string]interface{}{}))
+	if err != nil {
+		logger.Error("Unable to read search attributes from dynamic config. Search attributes migration is cancelled.", tag.Error(err))
+		return
+	}
+	dcCustomSearchAttributes := searchattribute.FilterCustomOnly(dcSearchAttributes)
+	if len(dcCustomSearchAttributes) == 0 {
+		logger.Debug("Search attributes are not defined in dynamic config. Search attributes migration is cancelled.", tag.Error(err))
+		return
+	}
+
+	factory := persistenceClient.NewFactory(
```

Comment:
also close this factory?

Reply:
Closing any manager created by the factory cause factory to close too. This sounds very wrong to me, I will fix it one day.